From: Brian May <bam@debian.org>
Date: Sat, 24 Jun 2017 17:05:58 +1000
Subject: Use locally installed documentation sources

---
 docs/conf.py | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/docs/conf.py b/docs/conf.py
index e531a8b40b89..510a550d01e3 100644
--- a/docs/conf.py
+++ b/docs/conf.py
@@ -177,9 +177,9 @@ pygments_style = "trac"
 # Links to Python's docs should reference the most recent version of the 3.x
 # branch, which is located at this URL.
 intersphinx_mapping = {
-    "python": ("https://docs.python.org/3/", None),
-    "sphinx": ("https://www.sphinx-doc.org/en/master/", None),
-    "psycopg": ("https://www.psycopg.org/psycopg3/docs/", None),
+    "python": ("/usr/share/doc/python3-doc/html/", None),
+    "sphinx": ("/usr/share/doc/sphinx-doc/html/", None),
+    "psycopg": ("/usr/share/doc/python-psycopg2-doc/html/", None),
 }
 
 # Python's docs don't change every week.
