From: Chris Lamb <lamby@debian.org>
Date: Tue, 26 Sep 2017 09:19:01 +0100
Subject: Use #!/usr/bin/env python3 shebang for django-admin.py.

---
 django/bin/django-admin.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/django/bin/django-admin.py b/django/bin/django-admin.py
index 594b0f11db53..898788e66b77 100755
--- a/django/bin/django-admin.py
+++ b/django/bin/django-admin.py
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python3
 # When the django-admin.py deprecation ends, remove this script.
 import warnings
 
