--- a/docs/_ext/djangodocs.py
+++ b/docs/_ext/djangodocs.py
@@ -10,7 +10,12 @@
 from sphinx import addnodes
 from sphinx.builders.html import StandaloneHTMLBuilder
 from sphinx.domains.std import Cmdoption
-from sphinx.util import logging
+
+try:  # sphinx>=1.6
+    from sphinx.util.logging import getLogger
+except ImportError:  # sphinx<1.6
+    from logging import getLogger
+
 from sphinx.util.console import bold
 from sphinx.util.nodes import set_source_info
 
@@ -19,7 +24,7 @@
 except ImportError:  # Sphinx 1.6+
     from sphinx.writers.html import HTMLTranslator
 
-logger = logging.getLogger(__name__)
+logger = getLogger(__name__)
 # RE for option descriptions without a '--' prefix
 simple_option_desc_re = re.compile(
     r'([-_a-zA-Z0-9]+)(\s*.*?)(?=,\s+(?:/|-|--)|$)')
