Description: Fix Django to build with Sphinx 1.1.2
Origin: upstream, https://code.djangoproject.com/changeset/16223

--- a/docs/_ext/djangodocs.py	(revision 14652)
+++ b/docs/_ext/djangodocs.py	(revision 16223)
@@ -128,4 +128,5 @@
     # Don't use border=1, which docutils does by default.
     def visit_table(self, node):
+        self._table_row_index = 0 # Needed by Sphinx
         self.body.append(self.starttag(node, 'table', CLASS='docutils'))
 
